File: 0146-lru-cache/0146-lru-cache.py
from dataclasses import dataclass

@dataclass
class Node:
    key: int
    val: int
    next: Node | None = None
    prev: Node | None = None

class LinkedList:
    def __init__(self):
        self._tail = Node(-1, -1)
        self._head = Node(-1, -1)
        self._tail.next = self._head
        self._head.prev= self._tail

    def add_front(self, node):
        self._head.prev.next = node
        node.prev = self._head.prev

        node.next = self._head
        self._head.prev = node

    def delete_node(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

        node.prev= None
        node.next = None

    def remove_last(self):
        lru_node = self._tail.next
        self._tail.next = lru_node.next
        lru_node.next.prev = self._tail

        lru_node.next = None
        lru_node.prev = None

        return lru_node


class LRUCache:

    def __init__(self, capacity: int):
        self._cap: int = capacity
        self._store: dict[int, Node] = {}
        self._ordering: LinkedList = LinkedList()

    def _key_accessed(self, key, node):
        self._ordering.delete_node(node)
        self._ordering.add_front(node)

    def _evict(self):
        return self._ordering.remove_last()
        

    def get(self, key: int) -> int:
        node = self._store.get(key)
        if node is not None:
            self._key_accessed(key, node)
            return self._store[key].val;
        return -1
        

    def put(self, key: int, value: int) -> None:
        node = self._store.get(key)
        if node is not None:
            node.val = value
            self._key_accessed(key, node)
        else:
            if len(self._store) == self._cap:
                lru_node = self._evict()
                del self._store[lru_node.key]


            node = Node(key, value)
            self._store[key] = node
            self._ordering.add_front(node)

        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
import java.util.HashMap;
import java.util.Map;

class LRUCache {

    // --- Inner Node Class for Doubly-Linked List ---
    class Node {
        int key;    // Must store the key to remove the entry from the HashMap upon eviction
        int val;
        Node prev;
        Node next;

        // Constructor for actual data nodes
        Node(int key, int val) {
            this.key = key;
            this.val = val;
        }

        // Constructor for dummy head/tail nodes (val is ignored)
        Node(int val) {
            this.val = val;
        }
    }

    // --- LRUCache Fields (Instance Variables) ---
    private final int capacity;
    // Maps the key to its corresponding Node in the linked list for O(1) lookups
    private final Map<Integer, Node> cache; 
    
    // Dummy head and tail nodes to simplify list operations (head.next is MRU, tail.prev is LRU)
    private final Node head;
    private final Node tail;

    // --- Constructor ---
    public LRUCache(int capacity) {
        // Fix 1: Initialize fields in the class, not in the constructor with 'static'.
        this.capacity = capacity;
        this.cache = new HashMap<>();
        
        this.head = new Node(-1);
        this.tail = new Node(-1);
        
        // Link head and tail initially
        head.next = tail;
        tail.prev = head;
    }

    // --- Helper Methods for Doubly-Linked List (O(1)) ---
    
    /**
     * Removes a given node from the list.
     */
    private void removeNode(Node n) {
        // Fix 2: The implementation you had for removeNode looked correct conceptually,
        // but it needs to be an instance method.
        n.prev.next = n.next;
        n.next.prev = n.prev;
        // Optionally clear references for Garbage Collection, though not strictly required
        n.prev = null;
        n.next = null;
    }

    /**
     * Adds a given node right after the head (making it the MRU).
     */
    private void addFront(Node n) {
        Node currentMRU = head.next;
        
        // 1. Link new node 'n'
        n.next = currentMRU;
        n.prev = head;
        
        // 2. Link surrounding nodes to 'n'
        head.next = n;
        currentMRU.prev = n;
    }

    // --- Main Methods ---

    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }

        Node node = cache.get(key);
        
        // Update Recency: move node to the front (MRU)
        removeNode(node);
        addFront(node);
        
        return node.val;
    }

    public void put(int key, int value) {
        // 1. Update Case
        if (cache.containsKey(key)) {
            Node node = cache.get(key);
            node.val = value; // Update the value
            
            // Update Recency: move node to the front (MRU)
            removeNode(node);
            addFront(node);
            return;
        }

        // 2. New Insertion Case (Check Capacity)
        if (cache.size() == capacity) {
            // Evict LRU item (the node just before the dummy tail)
            Node lruNode = tail.prev;
            
            // Remove from HashMap using the key stored in the LRU node
            cache.remove(lruNode.key);
            
            // Remove from Doubly-Linked List
            removeNode(lruNode);
        }

        // 3. Add New Entry
        Node newNode = new Node(key, value);
        
        // Add to HashMap and to the front of the list (MRU)
        cache.put(key, newNode);
        addFront(newNode);
    }
}